### PR TITLE
Add reader for FY-4B / GHI data

### PR DESCRIPTION
**Pull request by simonrp84**
This PR adds support for the Geostationary High-speed Imager aboard FengYun-4B. Overall it's quite similar to the existing AGRI reader but has some specific modifications to deal with the geolocation for GHI (which only samples a small portion of the full disk).

 - [x] Tests added
 - [x] Fully documented

### CODE REVIEW

**Comment by simonrp84** [comment]
pre-commit.ci autofix

**Comment by simonrp84** [comment]
pre-commit.ci autofix

**Comment by simonrp84** [comment]
pre-commit.ci autofix

**Comment by zxdawn** [comment]
The [FY4B document](http://img.nsmc.org.cn/PORTAL/NSMC/DATASERVICE/AuxiliaryData/FY4B/FY-4B%e6%a0%87%e7%a7%b0%e4%b8%8a%e8%a1%8c%e5%88%97%e5%8f%b7%e5%92%8c%e7%bb%8f%e7%ba%ac%e5%ba%a6%e7%9a%84%e4%ba%92%e7%9b%b8%e8%bd%ac%e6%8d%a2-133E.pdf) is available online now. The 250m COFF, CFAC, LOFF, and LFC need to be updated.

**Comment by simonrp84** [comment]
Thanks. Looks like my guesstimates were not too bad ;-)
I'll update the reader early next week.

**Comment by simonrp84** [comment]
pre-commit.ci autofix

**Comment by simonrp84** [comment]
pre-commit.ci autofix

**Comment by djhoese** [comment]
As mentioned on slack, after talking with @simonrp84 and Kathy Strabala I'm starting to think that it would be better to have two readers as @simonrp84 originally had it. Simon pointed out that a user doing:

```
scn = Scene(reader='agri_l1', filenames...)
scn.load(["C10"])
```

and having to know *what* "C10" they were requesting is too confusing. It should be more explicit than that. Despite what we would have liked as far as instrument naming and design (not changing channel properties for the same channel name), we have what we have and have to live with it. I think being more explicit with two separate readers (with names that match the reader naming conventions) is best. @mraspaud rebuttal?

**Comment by mraspaud** [comment]
I'm not so much against it actually. It's unfortunate, but maybe it's for the best. As long as the file patterns are clearly different to allow `find_fles_and_readers` to do its job properly, then I'm willing to see what you come up with :)

**Comment by mraspaud** [comment]
Any news on this?

File: satpy/composites/agri.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Copyright (c) 2015-2017 Satpy developers
```
Comment (djhoese) [inline satpy/composites/agri.py:3]: What's the actual year range for this file?

File: satpy/composites/agri.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+"""Composite classes for the AGRI instrument."""
+
+import logging
+
+from satpy.composites import GenericCompositor
+
+LOG = logging.getLogger(__name__)
+
+
+class PseudoRed_AGRI(GenericCompositor):
```
Comment (djhoese) [inline satpy/composites/agri.py:27]:
```suggestion
class PseudoRedAGRI(GenericCompositor):
```

**Review by djhoese — COMMENTED** [review]
There is a lot here so I'm not going to pretend I read all of it, but I had a few very small requests.

**Comment by simonrp84** [comment]
Thanks for the comments @djhoese. There certainly is a lot in here - turned into a far bigger job than expected.

Have updated the geolocation and the tests and have confirmed accuracy by comparing output images to coastlines, so hopefully all in order. I know the separation of AGRI on FY4A and FY4B isn't ideal, but can't think of a better solution.

File: satpy/composites/agri.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+"""Composite classes for the AGRI instrument."""
+
+import logging
+
+from satpy.composites import GenericCompositor
+
+LOG = logging.getLogger(__name__)
+
+
+class PseudoRedAGRI(GenericCompositor):
```
Comment (mraspaud) [inline satpy/composites/agri.py:27]: To be consistent, this should be named `SimulatedRed`

File: satpy/tests/reader_tests/test_agri_l1.py [hunk 4]
```diff
@@ -38,12 +37,20 @@
...
-    2000: (-5495021.206414789, 5493021.198696349, -5487021.175541028, 5495021.206414789),
-    4000: (-5494021.20255557, 5490021.187118688, -5478021.140808046, 5494021.20255557)
-}
+RESOLUTION_LIST = [500, 1000, 2000, 4000]
+
+AREA_EXTENTS_BY_RESOLUTION = {'FY4A': {
+    500: (-5496021.008869, 5495021.005046, -5493520.999312, 5496021.008869),
+    1000: (-5496021.076004, 5494021.068334, -5491021.05683, 5496021.076004),
+    2000: (-5496021.210274, 5492021.194837, -5486021.171682, 5496021.210274),
+    4000: (-5496021.210274, 5488021.1794, -5476021.13309,  5496021.210274)
```
Comment (mraspaud) [inline satpy/tests/reader_tests/test_agri_l1.py:46]: I understand you have looked at the geometric accuracy of these. Have they been checked against both satellites and multiple repeat cycles?

File: satpy/tests/reader_tests/test_agri_l1.py [hunk 5]
```diff
@@ -38,12 +37,20 @@
...
+},
+    'FY4B': {
+    500: (-5496021.008869, 5495021.005046, -5493520.999312, 5496021.008869),
+    1000: (-5496021.076004, 5494021.068334, -5491021.05683, 5496021.076004),
+    2000: (-5496021.210274, 5492021.194837, -5486021.171682, 5496021.210274),
+    4000: (-5496021.210274, 5488021.1794, -5476021.13309, 5496021.210274)
```
Comment (mraspaud) [inline satpy/tests/reader_tests/test_agri_l1.py:52]: I can't see any difference between fy4a and fy4b extents, should they be factorised?

**Review by mraspaud — COMMENTED** [review]
Thanks for adding these readers! I just have a few comments, but otherwise this looks good.

File: satpy/tests/reader_tests/test_agri_l1.py (hunk 4 above)
Comment (simonrp84) [reply satpy/tests/reader_tests/test_agri_l1.py:46]: I did, but having looked more closely I see some geolocation issues that I can't pin down. In addition, the different resolutions of AGRI data have different area extents (by several hundred meters) and I can't track that down either.

**Comment by simonrp84** [comment]
I haven't finished this, but have run out of time to work on it - can't guarantee I'll be able to come back to it for some time so hopefully someone else can pick it up. Am happy to supply data.

GHI seems pretty much OK. AGRI is causing issues as:
1) Different channels have different area extents, which is causing the `native` resampler to fail if `reduce_data=True`
2) The areas don't quite match coastlines, particularly for FY-4A/AGRI. I'm unsure if this is due to poor L1 data, poor area definition or both.

**Comment by simonrp84** [comment]
FWIW, my feeling is to edit the LOFF/COFF values supplied by CMA as I don't trust them. If we make those factors of each other (they're presently not) then the `native` resampler problem at least should go away and, hopefully, it would also improve geolocation.

**Review by mraspaud — APPROVED** [review]
LGTM

**Pull request merged**